ansible/roles/vm_set/library/vm_topology.py

```diff
@@ -403,7 +403,13 @@ def add_mgmt_port_to_docker(self, mgmt_bridge, mgmt_ip, mgmt_gw, mgmt_ipv6_addr=
         self.add_ip_to_docker_if(MGMT_PORT_NAME, mgmt_ip, mgmt_ipv6_addr=mgmt_ipv6_addr, mgmt_gw=mgmt_gw, mgmt_gw_v6=mgmt_gw_v6, api_server_pid=api_server_pid)
 
     def add_bp_port_to_docker(self, mgmt_ip, mgmt_ipv6):
-        self.add_br_if_to_docker(self.bp_bridge, PTF_BP_IF_TEMPLATE % self.vm_set_name, BP_PORT_NAME)
+        # add vm_set hash prefix to port name to support multiple tasks run concurrently
+        tmp_bp_if = hashlib.md5((PTF_NAME_TEMPLATE % self.vm_set_name).encode("utf-8")).hexdigest()[0:6] + BP_PORT_NAME
```

Comment:
I know why the PR testing failed.
PR testing needs to perform `refresh-dut`. In this case, the PTF docker is not removed. Function `self.add_br_if_to_docker` called at line 408 won't create new veth pair because the `ptf-vms6-1-b` interface already exists.

I reproduced this issue and got below vm_topology log:
```
2021-10-09 06:15:43,003 INFO root#417: === For veth pair, add ptf-vms6-1-b to bridge br-b-vms6-1, set ddd8c3backplane to PTF docker
2021-10-09 06:15:43,003 DEBUG root#953: *** CMD: ifconfig -a ptf-vms6-1-b, grep: None, attempt: 1
2021-10-09 06:15:43,007 DEBUG root#979: *** OUTPUT: 
{
  "cmd": "ifconfig -a ptf-vms6-1-b",
  "grep_cmd": null,
  "ret_code": 0,
  "stdout": [
    "ptf-vms6-1-b: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500",
    "        inet6 fe80::8842:deff:fe70:c710  prefixlen 64  scopeid 0x20<link>",
    "        ether 8a:42:de:70:c7:10  txqueuelen 1000  (Ethernet)",
    "        RX packets 599065  bytes 50313038 (50.3 MB)",
    "        RX errors 0  dropped 0  overruns 0  frame 0",
    "        TX packets 707334  bytes 65198582 (65.1 MB)",
    "        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0",
    ""
  ],
  "stderr": []
}
2021-10-09 06:15:43,007 DEBUG root#953: *** CMD: ifconfig -a ptf-vms6-1-b, grep: None, attempt: 2
2021-10-09 06:15:43,011 DEBUG root#979: *** OUTPUT: 
{
  "cmd": "ifconfig -a ptf-vms6-1-b",
  "grep_cmd": null,
  "ret_code": 0,
  "stdout": [
    "ptf-vms6-1-b: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500",
    "        inet6 fe80::8842:deff:fe70:c710  prefixlen 64  scopeid 0x20<link>",
    "        ether 8a:42:de:70:c7:10  txqueuelen 1000  (Ethernet)",
    "        RX packets 599065  bytes 50313038 (50.3 MB)",
    "        RX errors 0  dropped 0  overruns 0  frame 0",
    "        TX packets 707334  bytes 65198582 (65.1 MB)",
    "        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0",
    ""
  ],
  "stderr": []
}
2021-10-09 06:15:43,012 DEBUG root#953: *** CMD: ifconfig -a ptf-vms6-1-b, grep: None, attempt: 3
2021-10-09 06:15:43,015 DEBUG root#979: *** OUTPUT: 
{
  "cmd": "ifconfig -a ptf-vms6-1-b",
  "grep_cmd": null,
  "ret_code": 0,
  "stdout": [
    "ptf-vms6-1-b: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500",
    "        inet6 fe80::8842:deff:fe70:c710  prefixlen 64  scopeid 0x20<link>",
    "        ether 8a:42:de:70:c7:10  txqueuelen 1000  (Ethernet)",
    "        RX packets 599065  bytes 50313038 (50.3 MB)",
    "        RX errors 0  dropped 0  overruns 0  frame 0",
    "        TX packets 707334  bytes 65198582 (65.1 MB)",
    "        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0",
    ""
  ],
  "stderr": []
}
2021-10-09 06:15:43,015 DEBUG root#953: *** CMD: brctl show br-b-vms6-1, grep: None, attempt: 1
2021-10-09 06:15:43,018 DEBUG root#979: *** OUTPUT: 
{
  "cmd": "brctl show br-b-vms6-1",
  "grep_cmd": null,
  "ret_code": 0,
  "stdout": [
    "bridge name\tbridge id\t\tSTP enabled\tinterfaces",
    "br-b-vms6-1\t\t8000.26613e52b713\tno\t\tVM0100-back",
    "\t\t\t\t\t\t\tVM0101-back",
    "\t\t\t\t\t\t\tVM0102-back",
    "\t\t\t\t\t\t\tVM0103-back",
    "\t\t\t\t\t\t\tptf-vms6-1-b"
  ],
  "stderr": []
}
2021-10-09 06:15:43,018 DEBUG root#953: *** CMD: ip link set ptf-vms6-1-b up, grep: None, attempt: 1
2021-10-09 06:15:43,021 DEBUG root#979: *** OUTPUT: 
{
  "cmd": "ip link set ptf-vms6-1-b up",
  "grep_cmd": null,
  "ret_code": 0,
  "stdout": [],
  "stderr": []
}
2021-10-09 06:15:43,021 DEBUG root#953: *** CMD: ifconfig -a ddd8c3backplane, grep: None, attempt: 1
2021-10-09 06:15:43,028 DEBUG root#979: *** OUTPUT: 
{
  "cmd": "ifconfig -a ddd8c3backplane",
  "grep_cmd": null,
  "ret_code": 1,
  "stdout": [],
  "stderr": [
    "ddd8c3backplane: error fetching interface information: Device not found"
  ]
}
2021-10-09 06:15:43,028 DEBUG root#953: *** CMD: ifconfig -a ddd8c3backplane, grep: None, attempt: 2
2021-10-09 06:15:43,034 DEBUG root#979: *** OUTPUT: 
{
  "cmd": "ifconfig -a ddd8c3backplane",
  "grep_cmd": null,
  "ret_code": 1,
  "stdout": [],
  "stderr": [
    "ddd8c3backplane: error fetching interface information: Device not found"
  ]
}
2021-10-09 06:15:43,035 DEBUG root#953: *** CMD: ifconfig -a ddd8c3backplane, grep: None, attempt: 3
2021-10-09 06:15:43,041 DEBUG root#979: *** OUTPUT: 
{
  "cmd": "ifconfig -a ddd8c3backplane",
  "grep_cmd": null,
  "ret_code": 1,
  "stdout": [],
  "stderr": [
    "ddd8c3backplane: error fetching interface information: Device not found"
  ]
}
2021-10-09 06:15:43,041 DEBUG root#953: *** CMD: nsenter -t 580479 -n ip link set ddd8c3backplane up, grep: None, attempt: 1
2021-10-09 06:15:43,049 DEBUG root#979: *** OUTPUT: 
{
  "cmd": "nsenter -t 580479 -n ip link set ddd8c3backplane up",
  "grep_cmd": null,
  "ret_code": 1,
  "stdout": [],
  "stderr": [
    "Cannot find device \"ddd8c3backplane\""
  ]
}
2021-10-09 06:15:43,050 ERROR root#1425: Traceback (most recent call last):
  File "/tmp/ansible_vm_topology_payload_3i1ltdfy/__main__.py", line 1285, in main
    net.add_bp_port_to_docker(ptf_bp_ip_addr, ptf_bp_ipv6_addr)
  File "/tmp/ansible_vm_topology_payload_3i1ltdfy/__main__.py", line 408, in add_bp_port_to_docker
    self.add_br_if_to_docker(self.bp_bridge, PTF_BP_IF_TEMPLATE % self.vm_set_name, tmp_bp_if)
  File "/tmp/ansible_vm_topology_payload_3i1ltdfy/__main__.py", line 430, in add_br_if_to_docker
    VMTopology.iface_up(int_if, self.pid)
  File "/tmp/ansible_vm_topology_payload_3i1ltdfy/__main__.py", line 915, in iface_up
    return VMTopology.iface_updown(iface_name, 'up', pid)
  File "/tmp/ansible_vm_topology_payload_3i1ltdfy/__main__.py", line 926, in iface_updown
    return VMTopology.cmd('nsenter -t %s -n ip link set %s %s' % (pid, iface_name, state))
  File "/tmp/ansible_vm_topology_payload_3i1ltdfy/__main__.py", line 997, in cmd
    raise Exception('ret_code=%d, error message="%s". cmd="%s"' % (ret_code, err, cmdline))
Exception: ret_code=1, error message="Cannot find device "ddd8c3backplane"
". cmd="nsenter -t 580479 -n ip link set ddd8c3backplane up"
```
To address this issue, you need to wrap code of line 406-410 in `if` clause like:
```
        if VMTopology.intf_not_exists(BP_PORT_NAME, self.pid):
```
BTW, I have another concern. What if the first hash digit is a number, then the interface name would start with a number. Could such interface name be invalid? It may be more robust to generate the tmp_bp_if name like below:
```
        tmp_bp_if =  BP_PORT_NAME + hashlib.md5((PTF_NAME_TEMPLATE % self.vm_set_name).encode("utf-8")).hexdigest()[0:6]
```

It would be better to update the tmp_mgmt_if name as well.

Reply:
Thanks for the bug tracing,
I've modify the logic of add_br_if_to_docker, the function itself should provide such a multi-thread safe ability.
So now the suffix will be generate in add_br_if_to_docker, and outside caller just tell add_br_if_to_docker the target interface in ptf they want.
Will be much cleaner in the future when function add_br_if_to_docker is called other side.